File: src/main/java/eu/oakroot/TidySaveFieListener.java
package eu.oakroot;

import com.intellij.openapi.command.CommandProcessor;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManagerListener;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;


public class TidySaveFieListener implements FileDocumentManagerListener {
    private static final Logger LOG = Logger.getInstance(TidySaveFieListener.class);
    @Override
    public void beforeDocumentSaving(@NotNull Document document) {
        Project[] projects = ProjectManager.getInstance().getOpenProjects();
        for(Project project: projects) {
            if (!TidyImportsSettingsConfigurable.isOptionActive(project, TidyImportsOptionsForm.FORMAT_ON_SAVE)) {
                continue;
            }
            GoImportTidy goImportTidy = new GoImportTidy();

            String importsBlockStr = goImportTidy.findImports(document.getText());
            ArrayList<String> importsBlock = new ArrayList<>(Arrays.asList(importsBlockStr.split("\n")));
            String local = TidyImportsSettingsConfigurable.getOptionTextString(project, TidyImportsOptionsForm.LOCAL_PREFIX);
            try {
                ParsedFile parsedFile = goImportTidy.parseFile(importsBlock, local);
                if (parsedFile.isParsed()) {
                    CommandProcessor.getInstance().runUndoTransparentAction(new Runnable() {
                        @Override
                        public void run() {
                            document.setText(document.getText().replaceAll(importsBlockStr, parsedFile.getFileContent()));
                        }
                    });
                }
            } catch (IOException e) {
                LOG.debug(e.getMessage());
            }
        }
    }
}

File: src/main/kotlin/eu/oakroot/TidySaveFieListener.kt
package eu.oakroot

import com.intellij.openapi.command.CommandProcessor
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileDocumentManagerListener
import com.intellij.openapi.project.guessProjectForFile
import com.intellij.openapi.vfs.VirtualFile
import java.io.IOException
import java.util.*


class TidySaveFieListener : FileDocumentManagerListener {

    private val myDocumentsToStripLater: MutableSet<Document> = HashSet()

    override fun beforeAllDocumentsSaving() {
        val documentsToStrip: Set<Document> = HashSet(myDocumentsToStripLater)
        myDocumentsToStripLater.clear()
        for (document in documentsToStrip) {
            tidy(document)
        }
    }

    override fun beforeDocumentSaving(document: Document) {
        tidy(document)
    }

    private fun tidy(document: Document) {
        val file: VirtualFile = FileDocumentManager.getInstance().getFile(document) ?: return
        val project = guessProjectForFile(file)?: return
        val goImportTidy = GoImportTidy()
        val importsBlockStr = goImportTidy.findImports(document.text)
        val importsBlock =
            ArrayList(Arrays.asList(*importsBlockStr.split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()))
        val locals: List<String> = TidyImportsSettingsConfigurable.getOptionTextString(project, TidyImportsOptionsForm.LOCAL_PREFIX).split(",").map { it.trim() }.filter { it.isNotEmpty() }
        try {
            val parsedFile = goImportTidy.parseFile(importsBlock, locals)
            val importBlock = goImportTidy.extractBlockContent(document.text)
            if (parsedFile.isParsed) {
                CommandProcessor.getInstance().runUndoTransparentAction {
                    document.setText(
                        document.text.replace(
                            importBlock,
                            parsedFile.fileContent
                        )
                    )
                }
            }
        } catch (e: IOException) {
            LOG.debug(e.message)
        }
    }

    override fun unsavedDocumentsDropped() {
        myDocumentsToStripLater.clear()
    }

    companion object {
        private val LOG = Logger.getInstance(
            TidySaveFieListener::class.java
        )
    }
}